feat(linter/plugins): add `createOnce` API

overlookmotel opened the pull request:
Add `createOnce` method for rules. This is the "alternative API" mentioned in https://github.com/oxc-project/oxc/issues/9905#issuecomment-3208424180.

```js
export default {
  meta: { /* ... */ },
  // `createOnce`, not `create`. Called only once, NOT once per file.
  createOnce(context) {
    let classCount;
    return {
      before() {
        classCount = 0;
      },
      ClassDeclaration(klass) {
        classCount++;
      },
      after() {
        if (classCount > 5) {
          context.report({
            message: 'files should not contain more than 5 class declarations, '
              + `found ${classCount}`,
            node: { start: 0, end: 0 },
          });
        }
      },
    };
  }
};
```

If `before` method returns `false`, the rule is skipped for the current file. This is equivalent to the following pattern in ESLint:

```js
export default {
  meta: { /* ... */ },
  create(context) {
    if (shouldSkip(context)) return {}; // Skip
    return { /* visitor */ };
  }
};
```

Currently we don't capitalize on the potential perf benefits of this alternative API, but plugin authors can use it now, and once we do implement the perf optimizations this API enables, they'll get a speed boost just by upgrading Oxlint, without needing to change their plugin code.

camc314 reviewed (approved):
@overlookmotel  i haven't added the merge tag on these two because there are prs lower in the stack that're assigned to you.

but both of these look ok to me to merge

overlookmotel committed:
feat(linter/plugins): add `createOnce` API (#13944)

Add `createOnce` method for rules. This is the "alternative API" mentioned in https://github.com/oxc-project/oxc/issues/9905#issuecomment-3208424180.

```js
export default {
  meta: { /* ... */ },
  // `createOnce`, not `create`. Called only once, NOT once per file.
  createOnce(context) {
    let classCount;
    return {
      before() {
        classCount = 0;
      },
      ClassDeclaration(klass) {
        classCount++;
      },
      after() {
        if (classCount > 5) {
          context.report({
            message: 'files should not contain more than 5 class declarations, '
              + `found ${classCount}`,
            node: { start: 0, end: 0 },
          });
        }
      },
    };
  }
};
```

If `before` method returns `false`, the rule is skipped for the current file. This is equivalent to the following pattern in ESLint:

```js
export default {
  meta: { /* ... */ },
  create(context) {
    if (shouldSkip(context)) return {}; // Skip
    return { /* visitor */ };
  }
};
```

Currently we don't capitalize on the potential perf benefits of this alternative API, but plugin authors can use it now, and once we do implement the perf optimizations this API enables, they'll get a speed boost just by upgrading Oxlint, without needing to change their plugin code.
